File: app/src/main/java/ie/dublinbuspal/android/data/remote/logging/NetworkLoggingInterceptor.java
package ie.dublinbuspal.android.data.remote.logging;

import android.util.Log;

import java.io.IOException;

import androidx.annotation.NonNull;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class NetworkLoggingInterceptor implements Interceptor {

    private static final String TAG = "api.network.call";

    public NetworkLoggingInterceptor() {
        super();
    }

    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        Request request = chain.request();

        long t1 = System.nanoTime();
        Log.d(TAG, String.format("Sending request %s on %s%n%s",
                request.url(), chain.connection(), request.headers()));

        Response response = chain.proceed(request);

        long t2 = System.nanoTime();
        Log.d(TAG, String.format("Received response for %s in %.1fms%n%s",
                response.request().url(), (t2 - t1) / 1e6d, response.headers()));

        return response;
    }

}

File: service/src/main/kotlin/ie/dublinbuspal/service/interceptor/NetworkLoggingInterceptor.kt
package ie.dublinbuspal.service.interceptor

import okhttp3.Interceptor
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class NetworkLoggingInterceptor : Interceptor {

    private val httpLoggingInterceptor: Interceptor by lazy { newHttpLoggingInterceptor() }
    private val log: Logger by lazy { newLogger() }

    override fun intercept(chain: Interceptor.Chain): Response {
        return httpLoggingInterceptor.intercept(chain)
    }

    private fun newLogger(): Logger {
        return LoggerFactory.getLogger(javaClass)
    }

    private fun newHttpLoggingInterceptor(): Interceptor {
        val interceptor = HttpLoggingInterceptor(HttpLoggingInterceptor.Logger {
            log.debug(it)
        })
        interceptor.level = HttpLoggingInterceptor.Level.HEADERS
        return interceptor
    }

}
